app/models/llms/basic_workflow/integration.py

```diff
@@ -0,0 +1,391 @@
+        budget_found = False
+        for keyword, budget_value in budget_patterns.items():
+            if keyword in user_input_lower:
+                try:
+                    budget_list = [b for b in Budget if b.value == budget_value]
+                    if 'Luxury' in budget_list and 'Budget-friendly' in budget_list:
+                        preferences['budget'] = 'Moderate'
+                    else:
+                        preferences['budget'] = budget_list[0]
+                except StopIteration:
+                    preferences['budget'] = budget_value
+                budget_found = True
+                break
+        
+        if not budget_found:
+            preferences['budget'] = default_prefs['budget']
+        
+        # There are can be only one travel style for one trip, so we choose the first one that matches the user input
+        style_patterns = self.preferences_config["travel_style_patterns"]
+        style_found = False
+        for keyword, style_value in style_patterns.items():
+            if keyword in user_input_lower:
+                try:
+                    style_enum = next(s for s in TravelStyle if s.value == style_value)
+                    preferences['travel_style'] = style_enum.value
+                except StopIteration:
+                    preferences['travel_style'] = style_value
+                style_found = True
+                break
+        
+        if not style_found:
+            preferences['travel_style'] = default_prefs['travel_style']
+        
+        preferences['additional_context'] = user_input
+        
+        return preferences
+
+    def generate_enhanced_itinerary(self, user_input: str) -> str:
+        """
+        Generate an enhanced itinerary by parsing user input and creating a detailed prompt.
+        
+        Args:
+            user_input: Raw user input string
+            
+        Returns:
+            str: Generated itinerary text
+        """
+        with timer("enhanced itinerary generation"):
+            preferences = self.parse_travel_request(user_input)
+            
+            detailed_prompt = self._create_user_prompt(
+                destination=preferences['destination'],
+                duration=preferences['duration'],
+                interests=preferences['interests'],
+                budget=preferences['budget'],
+                travel_style=preferences['travel_style'],
+                additional_context=preferences['additional_context']
+            )
+            
+            return self.generate_itinerary(detailed_prompt)
+
+
+def save_to_session_history(session_history: list, user_input: str, itinerary: str, preferences: Dict[str, str]):
+    session_entry = {
+        'timestamp': datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
+        'user_input': user_input,
+        'preferences': preferences,
+        'itinerary': itinerary
+    }
+    session_history.append(session_entry)
+    return session_history
```

Comment:
The function appends to the session_history list in place, which could have unintended side effects if the list is shared. Return a new list instead of mutating the input list.
```
new_history = session_history + [session_entry]
return new_history
```